Optimize `DisjointSet`

gmou3 opened the pull request:
The methods `find` and `union` of the class `DisjointSet_of_hashables` pass through an unnecessary layer of input checks that slow down the code. I avoided these checks, cythonized more of the code, and edited the docstrings.

I can attest to the fact that these changes improve speed, and in a follow-up PR (#37839) I mention a test that confirms this.

tscrim commented on src/sage/sets/disjoint_set.pyx:
```diff
@@ -397,7 +397,7 @@ cdef class DisjointSet_of_integers(DisjointSet_class):
+        cdef list l = []
         cdef int i
         for i in range(self.cardinality()):
             l.append(self._nodes.parent[i])
```
```suggestion
        cdef Py_ssize_t card = self.cardinality()
        cdef list l = [None] * card
        cdef int i
        for i in range(card):
            l[i] = self._nodes.parent[i]
```
It is faster to set the length of the list than to constantly append. (Here it is a micro-optimization.)

tscrim commented on src/sage/sets/disjoint_set.pyx:
```diff
@@ -779,8 +826,44 @@ cdef class DisjointSet_of_hashables(DisjointSet_class):
-        r = self._d.find(i)
+        cdef int i = self._el_to_int[e]
+        cdef int r = self._d.find(i)
+        return self._int_to_el[r]
```
I don't think there is a benefit to typecasting here. It might actually be slower since it has to check the types.

tscrim commented on src/sage/sets/disjoint_set.pyx:
```diff
@@ -779,8 +826,44 @@ cdef class DisjointSet_of_hashables(DisjointSet_class):
+        """
+        cdef int i = self._el_to_int[e]
+        cdef int r = self._d._find(i)
         return self._int_to_el[r]
```
Same comment here about typecasting.

tscrim commented on src/sage/sets/disjoint_set.pyx:
```diff
@@ -797,23 +880,46 @@ cdef class DisjointSet_of_hashables(DisjointSet_class):
-        i = self._el_to_int[e]
-        j = self._el_to_int[f]
+        cdef int i = self._el_to_int[e]
+        cdef int j = self._el_to_int[f]
```
```suggestion
        cdef int i = <int> self._el_to_int[e]
        cdef int j = <int> self._el_to_int[f]
```
Avoid Cython checking the type.

tscrim commented on src/sage/sets/disjoint_set.pyx:
```diff
@@ -797,23 +880,46 @@ cdef class DisjointSet_of_hashables(DisjointSet_class):
+            {{'a', 'b', 'c', 'e'}, {'d'}}
+        """
+        cdef int i = self._el_to_int[e]
+        cdef int j = self._el_to_int[f]
```
```suggestion
        cdef int i = <int> self._el_to_int[e]
        cdef int j = <int> self._el_to_int[f]
```

tscrim reviewed (commented):
Here's some things to try and make it a bit faster.

gmou3 committed:
Use `cpdef`'s `union` and `find`

Typecast `union` as `void` to a-void the "C struct/union cannot be declared cpdef" error.

gmou3 committed:
Reintroduce cdef _union & _find

This avoids potential SEGFAULTs from Python calls

mantepse commented on src/sage/sets/disjoint_set.pyx:
```diff
@@ -379,28 +376,29 @@ cdef class DisjointSet_of_integers(DisjointSet_class):
-        for i in range(self.cardinality()):
-            l.append(self._nodes.parent[i])
+        for i in range(card):
+            l[i] = self._nodes.parent[i]
```
wouldn't `l = [self._nodes.parent[i] for i in range(self.cardinality())]` be better?

Hm, probably not: https://github.com/cython/cython/issues/2844

gmou3 replied:
Yes, I doubt it would help.

gmou3 committed:
Call `OP_find` and `OP_join` directly

From `find` and `union` of `DisjointSet_of_hashables`, respectively

gmou3 committed:
Various improvements

Remove `_find` & `_union`.
Remove `DisjointSet_of_integers` `_d` attribute from `DisjointSet_of_hashables` (and adjust code).
Add notes recommending `OP_find` and `OP_union` if no input checking needed.
Touch-up docstrings.

videlec commented on src/sage/sets/disjoint_set.pyx:
```diff
@@ -6,26 +6,25 @@ The main entry point is :func:`DisjointSet` which chooses the appropriate
 
-    sage: s = DisjointSet(6)
-    sage: s
+    sage: s = DisjointSet(6); s
```
It is better to avoid semi-colon to improve readability [see E702](https://www.flake8rules.com/rules/E702.html).

gmou3 replied:
Not so elegant, but ok.

videlec commented on src/sage/sets/disjoint_set.pyx:
```diff
@@ -379,28 +373,29 @@ cdef class DisjointSet_of_integers(DisjointSet_class):
             [0, 1, 0, 0, 4]
         """
-        l = []
+        cdef Py_ssize_t card = self.cardinality()
```
Why not use directly `self._nodes.degree` here?

videlec commented on src/sage/sets/disjoint_set.pyx:
```diff
@@ -531,57 +525,56 @@ cdef class DisjointSet_of_integers(DisjointSet_class):
+        cdef dict s = {}
+        cdef int i, o
         for i in range(self.cardinality()):
             o = self.find(i)
```
I thought it was faster to use `OP_find` :-)

gmou3 replied:
Indeed :O. I think these functions are not called as often as find and union.

videlec commented on src/sage/sets/disjoint_set.pyx:
```diff
@@ -531,57 +525,56 @@ cdef class DisjointSet_of_integers(DisjointSet_class):
+        cdef dict d = {}
         cdef int i
         for i in range(self.cardinality()):
             d[i] = self.find(i)
```
Again, use `OP_find`

videlec commented on src/sage/sets/disjoint_set.pyx:
```diff
@@ -531,57 +525,56 @@ cdef class DisjointSet_of_integers(DisjointSet_class):
-        d = {}
+        cdef dict d = {}
         cdef int i
         for i in range(self.cardinality()):
```
Again, use `self._nodes.degree`.

gmou3 committed:
Suggestions by videlec

gmou3 committed:
Merge branch 'develop' into disjointset_optimize

tscrim reviewed (approved):
I will take @videlec's comment to mean a positive review.
